### DISCUSSION TITLE
Installation on macOS just for analysis

### DISCUSSION THREAD

**Discussion by marsiwiec** [AllenInstitute/MIES#1851] [Q&A] [answered]
Hi,

I was wondering if it is possible to use the MIES module psx analysis (new feature in 2.7) for postsynaptic current detection on macOS, without hardware XOPs. I watched the YouTube video and it looks pretty great. I tried setting it up according to manual installation instructions but there were errors on startup.

Thanks,
Marcin

**Comment by t-b** [comment]
@marsiwiec Can you post the error message?

I bet it is due to the missing XOPs for MacOSX.

**Reply by marsiwiec** [reply]
@t-b Thanks, here is the message:

<img width="611" alt="image" src="https://github.com/AllenInstitute/MIES/assets/4427773/d1c02621-261c-4db5-8451-9309f24d4381">

This is what I get after running Igor once I download both MIES and the linked IPNWB folder and link it according to the installation instructions for manual installation without hardware dependencies: https://alleninstitute.github.io/MIES/installation.html

**Reply by t-b** [reply]
As I thougt. I'll bring that up in the next project meeting tomorrow.

**Reply by marsiwiec** [reply]
Thanks, I really appreciate it!

Marcin

**Reply by t-b** [reply]
A followup question: You are using MIES to acquire your data as well or? Or who have you planned to get your data into MIES?

**Reply by marsiwiec** [reply]
Right now I am using ACQ4 on a Sensapex rig (scope + manipulators + pressure controller), and I am really enjoying the semiautomatic patching capabilities. As an aside, do you have any workflows for combining approaching/patching cells with ACQ4 and subsequent recordings with MIES? I've noticed some scripts inside the ACQ4 repo but cannot figure out how to combine the two.

I would like to switch to MIES for executing certain recording protocols as I find it a lot more responsive and powerful especially for multipatch. The recordings are performed on a Windows machine, so the macOS MIES would only be involved in occasional PSC analysis.

I am also trying to import ABF files into MIES to analyse postsynaptic currents. Do you think a simple conversion to nwb2 using x-to-nwb or similar would be enough to be able to do the analysis?

**Reply by t-b** [reply]
> Right now I am using ACQ4 on a Sensapex rig (scope + manipulators + pressure controller), and I am really enjoying the semiautomatic patching capabilities. As an aside, do you have any workflows for combining approaching/patching cells with ACQ4 and subsequent recordings with MIES? [..]

@timjarsky will follow up on that one.

> I would like to switch to MIES for executing certain recording protocols as I find it a lot more responsive and powerful especially for multipatch. The recordings are performed on a Windows machine, so the macOS MIES would only be involved in occasional PSC analysis.

Nice to hear that you like MIES. I've created https://github.com/AllenInstitute/MIES/issues/1639 to track the addition of experimental MacOSX support without hardware.

> I am also trying to import ABF files into MIES to analyse postsynaptic currents. Do you think a simple conversion to nwb2 using x-to-nwb or similar would be enough to be able to do the analysis?

At the moment I don't think this is possible. I've created https://github.com/AllenInstitute/MIES/issues/1858 to track that.

**Reply by campagnola** [reply]
@marsiwiec the MIES bridge has only been used with an older version of ACQ4 (much older; python 2 only) and does not work with the latest. We will be working on this over the next several months, though, so stay tuned.

**Reply by marsiwiec** [reply]
@campagnola thanks for the info, I will stay tuned :)

**Comment by t-b** [comment]
@marsiwiec In https://github.com/AllenInstitute/MIES/pull/1859 I have prepared a first version of MIES with experimental MacOSX support. It passes the relevant part of the non-hardware tests and also includes all XOPs required for analysis. Can you give that a shot and report back if it works on your end?

**Comment by timjarsky** [comment]
@marsiwiec I've contacted @campagnola; he should be in touch soon about the ACQ4 code for MIES-ACQ4 integration (That's how we use it at the Allen Institute)

**Comment by marsiwiec** [comment]
@t-b Thanks! I will give this a try soon and report back.

@timjarsky Thank you. I look forward to integrating these tools. One more question - is there a way to analyze ACQ4 files in MIES? I have multisweep recordings of sIPSCs/mIPSCs done in ACQ4 and would love to analyze them in the new psc analysis module.

**Reply by t-b** [reply]
> One more question - is there a way to analyze ACQ4 files in MIES?

Not that I know of. I've created https://github.com/AllenInstitute/MIES/issues/1869 to track the request.

**Reply by marsiwiec** [reply]
Thanks. Maybe "dressing up" the ACQ4 dataset and exporting to NWB somehow would make it compatible? I honestly have no clear idea how to achieve that though. I think I will just try to use MIES for recording PSCs from now on. Planning to manually switch from patching a cell in ACQ4 (to make use of the semiautomatic capabilities) to recording protocols in MIES, as it is definitely superior in terms of data acquisition/analysis. 

One more question - I am trying to set everything up and it would really help if I had a sample config file for one of your rigs, just to see how to set everything up. Would that be ok?

**Reply by t-b** [reply]
I don't have access to the rig config files, but we do have some shipped with MIES, see https://github.com/AllenInstitute/MIES/tree/main/Packages/Settings_template.

**Comment by marsiwiec** [comment]
@t-b so I cloned the repo, switched to the 1859 branch, symlinked everything according to the documentation and although Igor starts I get this message in the console:

`  Experiment: Untitled ()
  Igor Pro version: 9.02 (39951)
  ################################
  Double threadsafe assertion Fail encountered !`

Also, there is no menu item for MIES.

**Reply by t-b** [reply]
Is that the full message?
Are the procedures compiled inside IP?
Do you have all MIES procedure files available or only MIES_Include? Can be checked by inspecting Windows->Procedure Windows.

**Reply by t-b** [reply]
What MacOSX version are you on?

**Reply by marsiwiec** [reply]
> What MacOSX version are you on?

Newest Ventura

> Is that the full message?
Are the procedures compiled inside IP?
Do you have all MIES procedure files available or only MIES_Include? Can be checked by inspecting Windows->Procedure Windows.

Yes, that's the full message. It seems like everything's compiled and present in Procedure Windows.

**Reply by t-b** [reply]
Okay, so that's probably some MacOSX security feature kicking in. I've added a commit to the PR which outputs some more information in the double threadsafe assertion fail case. Please try again and post the message.

**Reply by t-b** [reply]
Here is a screenshot from System Preferences->Security & Privacy->Files and Folders:

![image](https://github.com/AllenInstitute/MIES/assets/1185677/57341196-ca98-435c-8705-ec5e319bcfea)

Giving IP access here might be required for MIES to work. The screenshot is from Catalina.

**Reply by t-b** [reply]
@marsiwiec Did you get the chance to try the above steps?

**Reply by marsiwiec** [reply]
Hi @t-b, sorry for the late reply. I have checked the permissions, the menu looks a bit different on Sonoma and there is no option to add Igor to the list of apps under System Preferences->Security & Privacy->Files and Folders:

<img width="500" alt="Screenshot 2023-10-23 at 10 51 42" src="https://github.com/AllenInstitute/MIES/assets/4427773/b1305f7e-32d0-40f0-b322-46cc04ac25d1">

When I add the necessary shortcuts from the macOS fork now (both the ones from the XOPs-MacOSX-IP9-64bit folder and the remaining two from the XOPs-IP9-64bit folder), I get this error:

<img width="611" alt="Screenshot 2023-10-23 at 10 49 26" src="https://github.com/AllenInstitute/MIES/assets/4427773/ba956489-3408-4ef1-838c-bd6fd5673ee0">

**Reply by t-b** [reply]
I've pushed a fix for the compilation error.

>  (both the ones from the XOPs-MacOSX-IP9-64bit folder and the remaining two from the XOPs-IP9-64bit folder)

You don't need shortcuts to `XOPs-IP9-64bit` on MacOSX. The XOPs therein only work on Windows.

> there is no option to add Igor to the list of apps under System Preferences->Security & Privacy->Files and Folders

Hmm that's unfortunate.

Can you try to trigger the addition of IP into the list of known applications maybe with something like:
- Open IP
- Choose File->Open File->Procedure and select a ipf file inside Documents/WaveMetrics/...

But I'm only guessing here.

**Reply by marsiwiec** [reply]
> I've pushed a fix for the compilation error.

I do not see the fix, have you pushed it to the 1859 branch? I still get the same error after pulling newest changes.

> Can you try to trigger the addition of IP into the list of known applications maybe with something like:
>
> Open IP
> Choose File->Open File->Procedure and select a ipf file inside Documents/WaveMetrics/...

This does not trigger the dialog, so maybe it is whitelisted by default?

**Reply by t-b** [reply]
> I do not see the fix, have you pushed it to the 1859 branch? I still get the same error after pulling newest changes.

Oops! Just force pushed now.

> This does not trigger the dialog, so maybe it is whitelisted by default?

Hmm let's see that the assertion message is you are getting once it compiles on your end.

**Reply by marsiwiec** [reply]
Ok, thanks!

I had no compile error dialogs this time, but quite a bit of info about assertion failure in the console:

> !!! Assertion FAILED !!!
>   Message: "Expected a single letter for the drive"
>   Please provide the following information if you contact the MIES developers:
>   ################################
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   Stacktrace:
>   MEN_GetUserPingMenuString(...)#L240 [MIES_Menu.ipf]
>  GetUserPingEnabled(...)#L8447 [MIES_MiesUtilities.ipf]
>  GetSettingsJSONid(...)#L652 [MIES_GlobalStringAndVariableAccess.ipf]
>  PS_ReadSettings(...)#L41 [MIES_PackageSettings.ipf]
>  PS_GetSettingsFile(...)#L112 [MIES_PackageSettings.ipf]
>  PS_GetSettingsFolder(...)#L99 [MIES_PackageSettings.ipf]
>  PS_Initialize(...)#L20 [MIES_PackageSettings.ipf]
>  CreateFolderOnDisk(...)#L1823 [MIES_Utilities.ipf]
>  GetDrive(...)#L1804 [MIES_Utilities.ipf]
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   Time: 2023-10-25T21:16:42+02:00
>   Locked device: [- none -]
>   Current sweep: [- none -]
>   DAQ: [- none -]
>   Testpulse: [- none -]
>   Experiment: Untitled ()
>   Igor Pro version: 9.03B01 (56429)
>   MIES version:
>   
>   ################################
>   !!! Threadsafe assertion FAILED !!!
>   Message: "Missing initialization"
>   Please provide the following information if you contact the MIES developers:
>   ################################
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   Stacktrace:
>   MEN_GetUserPingMenuString(...)#L240 [MIES_Menu.ipf]
>  GetUserPingEnabled(...)#L8447 [MIES_MiesUtilities.ipf]
>  GetSettingsJSONid(...)#L652 [MIES_GlobalStringAndVariableAccess.ipf]
>  PS_ReadSettings(...)#L41 [MIES_PackageSettings.ipf]
>  PS_GetSettingsFile(...)#L112 [MIES_PackageSettings.ipf]
>  PS_GetSettingsFolder(...)#L99 [MIES_PackageSettings.ipf]
>  PS_Initialize(...)#L20 [MIES_PackageSettings.ipf]
>  CreateFolderOnDisk(...)#L1823 [MIES_Utilities.ipf]
>  GetDrive(...)#L1804 [MIES_Utilities.ipf]
>  ASSERT(...)#L187 [MIES_Utilities.ipf]
>  LOG_AddEntry(...)#L155 [MIES_Logging.ipf]
>  LOG_AddEntryWithoutChecks(...)#L76 [MIES_Logging.ipf]
>  LOG_GetFile(...)#L19 [MIES_Logging.ipf]
>  PS_GetSettingsFolder_TS(...)#L82 [MIES_PackageSettings.ipf]
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   Time: 2023-10-25T21:16:42+02:00
>   Experiment: Untitled ()
>   Igor Pro version: 9.03B01 (56429)
>   ################################
>   Double threadsafe assertion Fail encountered !
>   Missing initialization
>   MEN_GetUserPingMenuString,MIES_Menu.ipf,240;GetUserPingEnabled,MIES_MiesUtilities.ipf,8447;GetSettingsJSONid,MIES_GlobalStringAndVariableAccess.ipf,652;PS_ReadSettings,MIES_PackageSettings.ipf,41;
> PS_GetSettingsFile,MIES_PackageSettings.ipf,112;PS_GetSettingsFolder,MIES_PackageSettings.ipf,99;PS_Initialize,MIES_PackageSettings.ipf,20;CreateFolderOnDisk,MIES_Utilities.ipf,1823;
> GetDrive,MIES_Utilities.ipf,1804;ASSERT,MIES_Utilities.ipf,187;LOG_AddEntry,MIES_Logging.ipf,155;LOG_AddEntryWithoutChecks,MIES_Logging.ipf,76;LOG_GetFile,MIES_Logging.ipf,19;
> PS_GetSettingsFolder_TS,MIES_PackageSettings.ipf,82;ASSERT_TS,MIES_Utilities.ipf,266;LOG_AddEntry,MIES_Logging.ipf,155;LOG_AddEntryWithoutChecks,MIES_Logging.ipf,76;LOG_GetFile,MIES_Logging.ipf,19;
> PS_GetSettingsFolder_TS,MIES_PackageSettings.ipf,82;ASSERT_TS,MIES_Utilities.ipf,244;
>

**Reply by t-b** [reply]
Okay, that was helpful. I've pushed another fix, please update to 9f35ff17 (GetDrive: Make it work on Macintosh, 2023-10-25) on that branch.

**Reply by marsiwiec** [reply]
Ok, thanks, now it starts with no errors! Is there anything in particular you would like me to test for? I do not have sample traces recorded with MIES with me now, but might be able to get something next week.

**Reply by t-b** [reply]
> Ok, thanks, now it starts with no errors!

Very nice!

> Is there anything in particular you would like me to test for? I do not have sample traces recorded with MIES with me now, but might be able to get something next week.

I would not expect any big issues now that we got it so far. Do I'd say we merge the PR and you can then followup if something else comes up?

**Reply by marsiwiec** [reply, marked as answer]
> I would not expect any big issues now that we got it so far. Do I'd say we merge the PR and you can then followup if something else comes up?

Sounds great. I plan to start incorporating MIES into my daily recordings and gradually test the macOS analysis. Thanks!
